compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,445 @@
+        | Block (start, vars, is_array) -> (
+            match Var.Tbl.get live_table x with
+            | Live fields ->
+                let vars =
+                  Array.mapi
+                    ~f:(fun i v -> if IntSet.mem i fields then v else sentinal)
+                    vars
+                  |> compact_vars
+                in
+                let e = Block (start, vars, is_array) in
+                Let (x, e)
+            | _ -> instr)
+        | Apply ap ->
+            let args = List.map ~f:zero_var ap.args in
+            Let (x, Apply { ap with args })
+        | Field (_, _) | Closure (_, _) | Constant _ | Prim (_, _) -> instr)
+    | Assign (_, _) | Set_field (_, _, _) | Offset_ref (_, _) | Array_set (_, _, _) ->
+        instr
+  in
+  let zero_block block =
+    (* Analyze block instructions *)
+    let body = List.map ~f:(fun (instr, loc) -> zero_instr instr, loc) block.body in
+    (* Analyze branch *)
+    let branch =
+      (* Zero out return values in last instruction, otherwise do nothing. *)
+      match block.branch with
+      | Return x, loc -> Return (zero_var x), loc
+      | Raise (_, _), _
+      | Stop, _
+      | Branch _, _
+      | Cond (_, _, _), _
+      | Switch (_, _), _
+      | Pushtrap (_, _, _, _), _
+      | Poptrap _, _ -> block.branch
+    in
+    { block with body; branch }
+  in
+  let blocks = prog.blocks |> Addr.Map.map zero_block in
+  { prog with blocks }
+
+module Print = struct
+  let live_to_string = function
+    | Live fields ->
+        "live { " ^ IntSet.fold (fun i s -> s ^ Format.sprintf "%d " i) fields "" ^ "}"
+    | Top -> "top"
+    | Dead -> "dead"
+
+  let print_uses uses =
+    Format.eprintf "Usages:\n";
+    Var.Tbl.iter
+      (fun v ds ->
+        Format.eprintf "%a: { " Var.print v;
+        Var.Map.iter
+          (fun d k ->
+            Format.eprintf
+              "(%a, %s) "
+              Var.print
+              d
+              (match k with
+              | Compute -> "C"
+              | Propagate -> "P"))
+          ds;
+        Format.eprintf "}\n")
+      uses
+
+  let print_liveness live_vars =
+    Format.eprintf "Liveness:\n";
+    Var.Tbl.iter
+      (fun v l -> Format.eprintf "%a: %s\n" Var.print v (live_to_string l))
+      live_vars
+
+  let print_live_tbl live_table =
+    Format.eprintf "Liveness with dependencies:\n";
+    Var.Tbl.iter
+      (fun v l -> Format.eprintf "%a: %s\n" Var.print v (live_to_string l))
+      live_table
+end
+
+(** Add a sentinal variable declaration to the IR. The fresh variable is assigned to `undefined`. *)
+let add_sentinal p sentinal =
+  let undefined = Prim (Extern "%undefined", []) in
+  let instr, loc = Let (sentinal, undefined), noloc in
+  Code.prepend p [ instr, loc ]
+
+(** Run the liveness analysis and replace dead variables with the given sentinal. *)
+let f p ~deadcode_sentinal global_info =
+  let t = Timer.make () in
+  (* Add sentinal variable *)
+  let p = add_sentinal p deadcode_sentinal in
```

Comment:
Is this variable kept in the js file if unused ? I'm assuming deadcode won't be able to remove it because the external `%undefined"`is not known as pure. Can you maybe fix this and add a test ?

Reply:
I may be misunderstanding your question, but no? I tried this program:

```ocaml
(* testing/nodead.ml *)
let f x = x + 1
in print_int (f 1 + f 2 + f 3)
```
And ran this:
```
ocamlc -c testing/nodead.ml -o testing/nodead.cmo
dune exec -- js_of_ocaml --enable globaldeadcode --pretty testing/nodead.cmo
```
The result is:
```js
(function
  (globalThis){
   "use strict";
   var runtime = globalThis.jsoo_runtime;
   function caml_call1(f, a0){
    return (f.l >= 0 ? f.l : f.l = f.length) == 1
            ? f(a0)
            : runtime.caml_call_gen(f, [a0]);
   }
   var
    global_data = runtime.caml_get_global_data(),
    Stdlib = global_data.Stdlib;
   caml_call1(Stdlib[44], 9);
   var Nodead = [0];
   runtime.caml_register_global(1, Nodead, "Nodead");
   return;
  }
  (globalThis));
```

Reply:
It turns out that `undefined`, if unused, is removed by `Js_traverse.clean`. More precisely, by 

```ocaml
    method statements l =
      let l = super#statements l in
      List.filter l ~f:(function
          | (Empty_statement | Expression_statement (EVar _)), _ -> false
          | _ -> true)
```